snuba/cli/query_comparer.py

```diff
@@ -139,86 +53,186 @@ def query_comparer(
+        data_blob_exists = uploader.blob_exists(compared_data_blob)
+
+        if perf_blob_exists and data_blob_exists and not override:
             continue
 
-        assert v1_row[0] == v2_row[0], "Invalid query_ids: ids must match"
-        total_rows += 1
-
-        mismatches_exist = any(
-            [1 for i in range(1, len(v1_row)) if v1_row[i] != v2_row[i]]
-        )
-        if not mismatches_exist:
+        if compare_type == "perf" and perf_blob_exists and not override:
             continue
 
-        v1_data = query_measurement(v1_row)
-        v2_data = query_measurement(v2_row)
+        if compare_type == "data" and data_blob_exists and not override:
+            continue
 
-        mismatches.append(_create_mismatch(v1_data, v2_data))
+        result_file_format = file_manager.parse_blob_name(v1_blob)
+        v1_results = file_manager.download(v1_blob)
+        v2_results = file_manager.download(v2_blob)
+
+        perf_mismatches = []
+        data_mismatches = []
+        total_rows = 0
+        for v1_result, v2_result in itertools.zip_longest(v1_results, v2_results):
+
+            assert (
+                v1_result.query_id == v2_result.query_id
+            ), f"Invalid query_ids: {v1_result.query_id} and {v2_result.query_id} must match"
+            total_rows += 1
+            if v1_result == v2_result:
+                continue
+
+            assert isinstance(v1_result, QueryMeasurementResult)
+            assert isinstance(v2_result, QueryMeasurementResult)
+
+            perf_mismatch = create_perf_mismatch(v1_result, v2_result)
+            data_mismatch = create_data_mismatch(v1_result, v2_result)
+            if perf_mismatch:
+                perf_mismatches.append(perf_mismatch)
+            if data_mismatch:
+                data_mismatches.append(data_mismatch)
+
+        def save_comparisons(compare_type: str, mismatches: MismatchResultSet) -> None:
+            compared_file_format = FileFormat(
+                f"compared-{compare_type}",
+                result_file_format.date,
+                result_file_format.table,
+                result_file_format.hour,
+            )
+            file_manager.save(compared_file_format, mismatches, header_row=True)
+            # need the filename for the slack report
+            filename = file_manager.format_filename(compared_file_format)
+            _send_slack_report(
+                compare_type, total_rows, mismatches, result_file_format.table, filename
+            )
+
+        if compare_type == "perf":
+            save_comparisons(compare_type, perf_mismatches)
+        elif compare_type == "data":
+            save_comparisons(compare_type, data_mismatches)
+        else:
+            # no compare type == run both reports
+            save_comparisons("perf", perf_mismatches)
+            save_comparisons("data", data_mismatches)
+
+
+PERF_THRESHOLDS = {
+    "query_duration_ms": 0,
+    "read_rows": 0,
```

Comment:
Why have read_rows and read_bytes in perf thresholds?

Reply:
do we care to track if a query is reading more data than before? that would probably be reflected in the duration I guess

Reply:
for query performance, it shouldn't really matter as the first point of finding mismatches. you should only use query duration. If you want to drill down into mismatches, those additional fields may give us hints.